File: sicora-be-python/mevalservice/app/infrastructure/repositories/committee_repository.py
"""
Committee repository implementation.
"""

from datetime import date
from typing import Optional
from uuid import UUID

from sqlalchemy import and_, extract, select
from sqlalchemy.ext.asyncio import AsyncSession

from ...domain.entities import Committee, CommitteeStatus, CommitteeType
from ...domain.repositories import CommitteeRepository
from ..database.models import CommitteeModel, CommitteeStatusDB, CommitteeTypeDB


class CommitteeRepositoryImpl(CommitteeRepository):
    """SQLAlchemy implementation of CommitteeRepository."""

    def __init__(self, session: AsyncSession):
        self.session = session

    def _to_entity(self, model: CommitteeModel) -> Committee:
        """Convert database model to domain entity."""
        return Committee(
            id=model.id,
            committee_type=CommitteeType(model.committee_type.value),
            status=CommitteeStatus(model.status.value),
            scheduled_date=model.scheduled_date,
            actual_start_time=model.actual_start_time,
            actual_end_time=model.actual_end_time,
            location=model.location,
            is_virtual=model.is_virtual,
            virtual_meeting_link=model.virtual_meeting_link,
            training_center_id=model.training_center_id,
            agenda=model.agenda,
            minutes=model.minutes,
            minutes_approved_at=model.minutes_approved_at,
            minutes_approved_by=model.minutes_approved_by,
            created_by=model.created_by,
            created_at=model.created_at,
            updated_at=model.updated_at,
        )

    def _to_model(self, entity: Committee) -> CommitteeModel:
        """Convert domain entity to database model."""
        return CommitteeModel(
            id=entity.id,
            committee_type=CommitteeTypeDB(entity.committee_type.value),
            status=CommitteeStatusDB(entity.status.value),
            scheduled_date=entity.scheduled_date,
            actual_start_time=entity.actual_start_time,
            actual_end_time=entity.actual_end_time,
            location=entity.location,
            is_virtual=entity.is_virtual,
            virtual_meeting_link=entity.virtual_meeting_link,
            training_center_id=entity.training_center_id,
            agenda=entity.agenda,
            minutes=entity.minutes,
            minutes_approved_at=entity.minutes_approved_at,
            minutes_approved_by=entity.minutes_approved_by,
            created_by=entity.created_by,
            created_at=entity.created_at,
            updated_at=entity.updated_at,
        )

    async def create(self, committee: Committee) -> Committee:
        model = self._to_model(committee)
        self.session.add(model)
        await self.session.commit()
        await self.session.refresh(model)
        return self._to_entity(model)

    async def get_by_id(self, committee_id: UUID) -> Optional[Committee]:
        result = await self.session.execute(
            select(CommitteeModel).where(CommitteeModel.id == committee_id)
        )
        model = result.scalar_one_or_none()
        return self._to_entity(model) if model else None

    async def update(self, committee: Committee) -> Committee:
        result = await self.session.execute(
            select(CommitteeModel).where(CommitteeModel.id == committee.id)
        )
        model = result.scalar_one_or_none()
        if model:
            model.committee_type = CommitteeTypeDB(committee.committee_type.value)
            model.status = CommitteeStatusDB(committee.status.value)
            model.scheduled_date = committee.scheduled_date
            model.actual_start_time = committee.actual_start_time
            model.actual_end_time = committee.actual_end_time
            model.location = committee.location
            model.is_virtual = committee.is_virtual
            model.virtual_meeting_link = committee.virtual_meeting_link
            model.agenda = committee.agenda
            model.minutes = committee.minutes
            model.minutes_approved_at = committee.minutes_approved_at
            model.minutes_approved_by = committee.minutes_approved_by
            await self.session.commit()
            await self.session.refresh(model)
            return self._to_entity(model)
        raise ValueError(f"Committee {committee.id} not found")

    async def delete(self, committee_id: UUID) -> bool:
        result = await self.session.execute(
            select(CommitteeModel).where(CommitteeModel.id == committee_id)
        )
        model = result.scalar_one_or_none()
        if model:
            await self.session.delete(model)
            await self.session.commit()
            return True
        return False

    async def get_all(self, skip: int = 0, limit: int = 100) -> list[Committee]:
        result = await self.session.execute(
            select(CommitteeModel).offset(skip).limit(limit)
        )
        return [self._to_entity(m) for m in result.scalars().all()]

    async def get_by_date_range(
        self, start_date: date, end_date: date
    ) -> list[Committee]:
        result = await self.session.execute(
            select(CommitteeModel).where(
                and_(
                    CommitteeModel.scheduled_date >= start_date,
                    CommitteeModel.scheduled_date <= end_date,
                )
            )
        )
        return [self._to_entity(m) for m in result.scalars().all()]

    async def get_by_type(self, committee_type: CommitteeType) -> list[Committee]:
        result = await self.session.execute(
            select(CommitteeModel).where(
                CommitteeModel.committee_type == CommitteeTypeDB(committee_type.value)
            )
        )
        return [self._to_entity(m) for m in result.scalars().all()]

    async def get_by_status(self, status: CommitteeStatus) -> list[Committee]:
        result = await self.session.execute(
            select(CommitteeModel).where(
                CommitteeModel.status == CommitteeStatusDB(status.value)
            )
        )
        return [self._to_entity(m) for m in result.scalars().all()]

    async def get_by_training_center(self, training_center_id: UUID) -> list[Committee]:
        result = await self.session.execute(
            select(CommitteeModel).where(
                CommitteeModel.training_center_id == training_center_id
            )
        )
        return [self._to_entity(m) for m in result.scalars().all()]

    async def get_monthly_for_date(
        self, year: int, month: int, training_center_id: UUID
    ) -> Optional[Committee]:
        result = await self.session.execute(
            select(CommitteeModel).where(
                and_(
                    extract("year", CommitteeModel.scheduled_date) == year,
                    extract("month", CommitteeModel.scheduled_date) == month,
                    CommitteeModel.training_center_id == training_center_id,
                    CommitteeModel.committee_type == CommitteeTypeDB.MONTHLY,
                )
            )
        )
        model = result.scalar_one_or_none()
        return self._to_entity(model) if model else None

    async def get_next_scheduled(self, training_center_id: UUID) -> Optional[Committee]:
        result = await self.session.execute(
            select(CommitteeModel)
            .where(
                and_(
                    CommitteeModel.training_center_id == training_center_id,
                    CommitteeModel.status == CommitteeStatusDB.SCHEDULED,
                    CommitteeModel.scheduled_date >= date.today(),
                )
            )
            .order_by(CommitteeModel.scheduled_date.asc())
            .limit(1)
        )
        model = result.scalar_one_or_none()
        return self._to_entity(model) if model else None

File: sicora-be-go/mevalservice/internal/infrastructure/repositories/committee_repository.go
package repositories

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"

	"mevalservice/internal/domain/entities"
	"mevalservice/internal/domain/repositories"
	"mevalservice/internal/infrastructure/database"
)

type committeeRepository struct {
	db *gorm.DB
}

func NewCommitteeRepository(db *database.Database) repositories.CommitteeRepository {
	return &committeeRepository{
		db: db.DB,
	}
}

func (r *committeeRepository) Create(ctx context.Context, committee *entities.Committee) error {
	model := r.toModel(committee)
	if err := r.db.WithContext(ctx).Create(model).Error; err != nil {
		return err
	}
	committee.ID = model.ID
	return nil
}

func (r *committeeRepository) GetByID(ctx context.Context, id uuid.UUID) (*entities.Committee, error) {
	var model database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		First(&model, "id = ?", id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return r.toEntity(&model), nil
}

func (r *committeeRepository) GetAll(ctx context.Context, limit, offset int) ([]*entities.Committee, error) {
	var models []database.CommitteeModel
	query := r.db.WithContext(ctx).Preload("Members")
	if limit > 0 {
		query = query.Limit(limit).Offset(offset)
	}
	if err := query.Find(&models).Error; err != nil {
		return nil, err
	}

	committees := make([]*entities.Committee, len(models))
	for i, model := range models {
		committees[i] = r.toEntity(&model)
	}
	return committees, nil
}

func (r *committeeRepository) GetByType(ctx context.Context, committeeType entities.CommitteeType) ([]*entities.Committee, error) {
	var models []database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("committee_type = ?", string(committeeType)).
		Find(&models).Error; err != nil {
		return nil, err
	}

	committees := make([]*entities.Committee, len(models))
	for i, model := range models {
		committees[i] = r.toEntity(&model)
	}
	return committees, nil
}

func (r *committeeRepository) GetByStatus(ctx context.Context, status entities.CommitteeStatus) ([]*entities.Committee, error) {
	var models []database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("status = ?", string(status)).
		Find(&models).Error; err != nil {
		return nil, err
	}

	committees := make([]*entities.Committee, len(models))
	for i, model := range models {
		committees[i] = r.toEntity(&model)
	}
	return committees, nil
}

func (r *committeeRepository) GetByDateRange(ctx context.Context, startDate, endDate time.Time) ([]*entities.Committee, error) {
	var models []database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("committee_date BETWEEN ? AND ?", startDate, endDate).
		Find(&models).Error; err != nil {
		return nil, err
	}

	committees := make([]*entities.Committee, len(models))
	for i, model := range models {
		committees[i] = r.toEntity(&model)
	}
	return committees, nil
}

func (r *committeeRepository) GetByAcademicPeriod(ctx context.Context, academicPeriod string) ([]*entities.Committee, error) {
	var models []database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("academic_period = ?", academicPeriod).
		Find(&models).Error; err != nil {
		return nil, err
	}

	committees := make([]*entities.Committee, len(models))
	for i, model := range models {
		committees[i] = r.toEntity(&model)
	}
	return committees, nil
}

func (r *committeeRepository) GetByProgramID(ctx context.Context, programID uuid.UUID) ([]*entities.Committee, error) {
	var models []database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("program_id = ?", programID).
		Find(&models).Error; err != nil {
		return nil, err
	}

	committees := make([]*entities.Committee, len(models))
	for i, model := range models {
		committees[i] = r.toEntity(&model)
	}
	return committees, nil
}

func (r *committeeRepository) GetScheduledCommittees(ctx context.Context) ([]*entities.Committee, error) {
	var models []database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("status = ?", string(entities.CommitteeStatusProgramado)).
		Find(&models).Error; err != nil {
		return nil, err
	}

	committees := make([]*entities.Committee, len(models))
	for i, model := range models {
		committees[i] = r.toEntity(&model)
	}
	return committees, nil
}

func (r *committeeRepository) GetCompletedCommittees(ctx context.Context) ([]*entities.Committee, error) {
	var models []database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("status = ?", string(entities.CommitteeStatusCompletado)).
		Find(&models).Error; err != nil {
		return nil, err
	}

	committees := make([]*entities.Committee, len(models))
	for i, model := range models {
		committees[i] = r.toEntity(&model)
	}
	return committees, nil
}

func (r *committeeRepository) GetMonthlyCommitteeForDate(ctx context.Context, date time.Time) (*entities.Committee, error) {
	var model database.CommitteeModel
	startOfMonth := time.Date(date.Year(), date.Month(), 1, 0, 0, 0, 0, date.Location())
	endOfMonth := startOfMonth.AddDate(0, 1, -1)
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("committee_date BETWEEN ? AND ?", startOfMonth, endOfMonth).
		First(&model).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return r.toEntity(&model), nil
}

func (r *committeeRepository) GetNextScheduledCommittee(ctx context.Context) (*entities.Committee, error) {
	var model database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Where("status = ? AND committee_date > ?", string(entities.CommitteeStatusProgramado), time.Now()).
		Order("committee_date ASC").
		First(&model).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return r.toEntity(&model), nil
}

func (r *committeeRepository) GetCurrentMonthCommittees(ctx context.Context) ([]*entities.Committee, error) {
	now := time.Now()
	startOfMonth := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, now.Location())
	endOfMonth := startOfMonth.AddDate(0, 1, -1)
	return r.GetByDateRange(ctx, startOfMonth, endOfMonth)
}

func (r *committeeRepository) GetWithMembers(ctx context.Context, id uuid.UUID) (*entities.Committee, error) {
	return r.GetByID(ctx, id) // Already loads members via preload
}

func (r *committeeRepository) GetWithCases(ctx context.Context, id uuid.UUID) (*entities.Committee, error) {
	var model database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Preload("Cases").
		First(&model, "id = ?", id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return r.toEntity(&model), nil
}

func (r *committeeRepository) GetWithDecisions(ctx context.Context, id uuid.UUID) (*entities.Committee, error) {
	var model database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Preload("Decisions").
		First(&model, "id = ?", id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return r.toEntity(&model), nil
}

func (r *committeeRepository) GetWithAllRelations(ctx context.Context, id uuid.UUID) (*entities.Committee, error) {
	var model database.CommitteeModel
	if err := r.db.WithContext(ctx).
		Preload("Members").
		Preload("Cases").
		Preload("Decisions").
		First(&model, "id = ?", id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return r.toEntity(&model), nil
}

func (r *committeeRepository) Update(ctx context.Context, committee *entities.Committee) error {
	model := r.toModel(committee)
	return r.db.WithContext(ctx).Save(model).Error
}

func (r *committeeRepository) Delete(ctx context.Context, id uuid.UUID) error {
	return r.db.WithContext(ctx).Delete(&database.CommitteeModel{}, "id = ?", id).Error
}

func (r *committeeRepository) GetCommitteeCount(ctx context.Context) (int64, error) {
	var count int64
	if err := r.db.WithContext(ctx).Model(&database.CommitteeModel{}).Count(&count).Error; err != nil {
		return 0, err
	}
	return count, nil
}

func (r *committeeRepository) GetCommitteeCountByType(ctx context.Context, committeeType entities.CommitteeType) (int64, error) {
	var count int64
	if err := r.db.WithContext(ctx).Model(&database.CommitteeModel{}).
		Where("committee_type = ?", string(committeeType)).
		Count(&count).Error; err != nil {
		return 0, err
	}
	return count, nil
}

func (r *committeeRepository) GetCommitteeCountByDateRange(ctx context.Context, startDate, endDate time.Time) (int64, error) {
	var count int64
	if err := r.db.WithContext(ctx).Model(&database.CommitteeModel{}).
		Where("committee_date BETWEEN ? AND ?", startDate, endDate).
		Count(&count).Error; err != nil {
		return 0, err
	}
	return count, nil
}

// Conversion methods
func (r *committeeRepository) toModel(committee *entities.Committee) *database.CommitteeModel {
	return &database.CommitteeModel{
		ID:              committee.ID,
		CommitteeDate:   committee.CommitteeDate,
		CommitteeType:   string(committee.CommitteeType),
		Status:          string(committee.Status),
		ProgramID:       committee.ProgramID,
		AcademicPeriod:  committee.AcademicPeriod,
		AgendaGenerated: committee.AgendaGenerated,
		QuorumAchieved:  committee.QuorumAchieved,
		SessionMinutes:  committee.SessionMinutes,
		CreatedAt:       committee.CreatedAt,
		UpdatedAt:       committee.UpdatedAt,
	}
}

func (r *committeeRepository) toEntity(model *database.CommitteeModel) *entities.Committee {
	committee := &entities.Committee{
		ID:              model.ID,
		CommitteeDate:   model.CommitteeDate,
		CommitteeType:   entities.CommitteeType(model.CommitteeType),
		Status:          entities.CommitteeStatus(model.Status),
		ProgramID:       model.ProgramID,
		AcademicPeriod:  model.AcademicPeriod,
		AgendaGenerated: model.AgendaGenerated,
		QuorumAchieved:  model.QuorumAchieved,
		SessionMinutes:  model.SessionMinutes,
		CreatedAt:       model.CreatedAt,
		UpdatedAt:       model.UpdatedAt,
	}

	// Convert members if loaded
	if len(model.Members) > 0 {
		members := make([]entities.CommitteeMember, len(model.Members))
		for i, memberModel := range model.Members {
			members[i] = entities.CommitteeMember{
				ID:          memberModel.ID,
				CommitteeID: memberModel.CommitteeID,
				UserID:      memberModel.UserID,
				MemberRole:  entities.MemberRole(memberModel.MemberRole),
				IsPresent:   memberModel.Attended,
				VotePower:   1, // Default
				CreatedAt:   memberModel.CreatedAt,
			}
		}
		committee.Members = members
	}

	return committee
}
